flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/TaskBalancedPreferredSlotSharingStrategy.java

```diff
@@ -0,0 +1,270 @@
+                                    eSsgs.add(i, executionSlotSharingGroup);
+                                    LOG.debug(
+                                            "Set resourceProfile {} for {}th executionSlotSharingGroup {} in slotSharingGroup {}.",
+                                            ssg.getResourceProfile(),
+                                            i,
+                                            executionSlotSharingGroup,
+                                            ssg);
+                                }
+                                ssgToExecutionSSGs.put(ssg, eSsgs);
+                            });
+        }
+
+        /** Find the execution slot sharing group that holds the least execution tasks. */
+        private int findAvailableESSGForCoLocation(
+                SchedulingExecutionVertex sev, List<ExecutionSlotSharingGroup> executionSSGs) {
+            int indexWithLeastExecutionVertices = 0;
+            int leastExecutionVertices = Integer.MAX_VALUE;
+            for (int index = 0; index < executionSSGs.size(); index++) {
+                if (executionSSGs.get(index).getExecutionVertexIds().contains(sev.getId())) {
+                    continue;
+                }
+                if (leastExecutionVertices
+                        > executionSSGs.get(index).getExecutionVertexIds().size()) {
+                    indexWithLeastExecutionVertices = index;
+                    leastExecutionVertices =
+                            executionSSGs.get(index).getExecutionVertexIds().size();
+                }
+            }
+            return indexWithLeastExecutionVertices;
+        }
+
+        private CoLocationConstraint getCoLocationConstraint(
+                JobVertexID jobVertexID, SchedulingExecutionVertex sev) {
+            Preconditions.checkState(coLocationGroupMap.containsKey(sev.getId().getJobVertexId()));
+            return coLocationGroupMap
+                    .get(jobVertexID)
+                    .getLocationConstraint(sev.getId().getSubtaskIndex());
+        }
+
+        /**
+         * Add the SchedulingExecutionVertex into a ExecutionSlotSharingGroup by the index and
+         * return the next available index.
+         *
+         * @param executionSSGs All execution slot sharing groups for current SlotSharingGroup.
+         * @param index The index of the target ExecutionSlotSharingGroup.
+         * @param sev The target ExecutionSlotSharingGroup
+         * @return The next available round-robin slot index.
+         */
+        private int addVertexToExecutionSlotSharingGroup(
+                List<ExecutionSlotSharingGroup> executionSSGs,
+                int index,
+                SchedulingExecutionVertex sev) {
+            executionSSGs.get(index).addVertex(sev.getId());
+            executionSlotSharingGroupMap.put(sev.getId(), executionSSGs.get(index));
+            index = ++index % executionSSGs.size();
+            return index;
+        }
```

Comment:
Can it be simplified to this?

```
       private void addVertexToExecutionSlotSharingGroup(
                ExecutionSlotSharingGroup executionSlotSharingGroup,
                List<ExecutionSlotSharingGroup> executionSSGs,
                ExecutionVertexID executionVertexID) {
            executionSlotSharingGroup.addVertex(executionVertexID);
            executionSlotSharingGroupMap.put(executionVertexID, executionSlotSharingGroup);
        }
```

Reasons:
1. Don't care about index, it's clearer.
2. ` if (clcToESsgMap.get(clc) != null) {` of `TaskBalancedExecutionSlotSharingGroupBuilder#build` can use new `addVertexToExecutionSlotSharingGroup` as well.